lib/google/gax/bundling.rb

```diff
@@ -0,0 +1,456 @@
+      def event_for(elts)
+        event = Event.new
+        event.canceller = canceller_for(elts, event)
+        event
+      end
+
+      # Creates a cancellation proc that removes elts.
+      #
+      # The returned proc returns true if all elements were successfully removed
+      # from @in_deque and  @event_deque.
+      #
+      # @param elts [Array<Object>] an array of elements that can be appended
+      #     to the tasks bundle_field.
+      # @param [Event] an Event that can be used to wait on the response.
+      # @return [Proc] the canceller that when called removes the elts
+      #     and events.
+      def canceller_for(elts, event)
+        proc do
+          event_index = @event_deque.find_index(event) || -1
+          in_index = @in_deque.find_index(elts) || -1
+          @event_deque.delete_at(event_index) unless event_index == -1
+          @in_deque.delete_at(in_index) unless in_index == -1
+          if event_index == -1 || in_index == -1
+            false
+          else
+            true
+          end
+        end
+      end
+
+      # private_class_method :run_with_no_subresponse,
+      #                      :run_with_subresponses,
+      #                      :event_for, :cancellor_for
+    end
+
+    # Organizes bundling for an api service that requires it.
+    class Executor
+      # @param [BundleOptions]configures strategy this instance
+      #     uses when executing bundled functions.
+      def initialize(options, timer: Timer.new)
+        @options = options
+        @tasks = {}
+        @timer = timer
+
+        # Use a Monitor in order to have the mutex behave like a reentrant lock.
+        @tasks_lock = Monitor.new
+      end
+
+      # Schedules bundle_desc of bundling_request as part of
+      # bundle id.
+      #
+      # @param api_call [Proc] used to make an api call when the task is run.
+      # @param bundle_id [String] the id of this bundle.
+      # @param bundle_desc [BundleDescriptor] describes the structure of the
+      #     bundled call.
+      # @param bundling_request [Object] the request to pass as the arg to
+      #     the api_call.
+      # @param kwargs [Hash] the keyword arguments passed to the api_call.
+      # @return [Event] an Event that can be used to wait on the response.
+      def schedule(api_call, bundle_id, bundle_desc,
+                   bundling_request, kwargs: {})
+        bundle = bundle_for(api_call, bundle_id, bundle_desc,
+                            bundling_request, kwargs)
+        elts = bundling_request.send(bundle_desc.bundled_field.to_s)
+        event = bundle.extend(elts)
+
+        count_threshold = @options.element_count_threshold
+        if count_threshold > 0 && bundle.element_count >= count_threshold
+          run_now(bundle.bundle_id)
+        end
+
+        size_threshold = @options.request_byte_threshold
+        if size_threshold > 0 && bundle.request_bytesize >= size_threshold
+          run_now(bundle.bundle_id)
+        end
```

Comment:
Recommend to write a TODO comment for 'limits'.

For the background, when I was doing bundling for gax-nodejs, @geigerj pointed out that bundling option also has the parameter for 'limit', which GAX Python didn't implement yet. See: https://github.com/googleapis/gax-ruby/blob/master/lib/google/gax.rb#L264

However, from my experience of implementing it for NodeJS, supporting limit is actually more complex than it seems, so I recommend not doing it in this PR -- it's worth doing in another one (it also requires more test cases).  But it's better to leave a TODO comment.